saffron/src/lib.rs

```diff
@@ -1359,194 +1360,128 @@ impl Cron {
+    fn find_next_month(&self, date: Date<Utc>) -> Option<Date<Utc>> {
         let Months(map) = self.months;
+        let current_month = date.month0();
         let bottom_cleared = (map >> current_month) << current_month;
-        let trailing_zeroes = bottom_cleared.trailing_zeros();
-        if trailing_zeroes < Months::BITS as u32 {
-            Some(trailing_zeroes)
+        let trailing_zeros = bottom_cleared.trailing_zeros();
+        if trailing_zeros < Months::BITS as u32 {
+            Utc.ymd_opt(date.year(), trailing_zeros + 1, 1).single()
         } else {
             None
         }
     }
 
-    fn find_next_time(&self, time: NaiveTime) -> Option<NaiveTime> {
-        let hour = time.hour();
-        if self.hours.contains_hour(time) {
-            if let Some(next_minute) = self.find_next_minute(time.minute()) {
-                return Some(NaiveTime::from_hms(hour, next_minute, 0));
+    fn find_next_date(
+        &self,
+        mut date: Date<Utc>,
+        bound: Date<Utc>,
+    ) -> Result<Option<Date<Utc>>, OutOfBound> {
+        if self.months.contains_month(date) {
+            match self.find_next_day(date) {
+                Some(next_day) if next_day > bound => return Err(OutOfBound),
+                Some(next_day) => return Ok(Some(next_day)),
+                None => {}
             }
         }
 
-        if hour < 23 {
-            if let Some(next_hour) = self.find_next_hour(hour + 1) {
-                let minute = self.find_next_minute(0).expect(Self::MUST_CONTAIN_MINUTE);
+        loop {
+            date = match next_month_in_year(date) {
+                Some(next_month) if next_month > bound => return Err(OutOfBound),
+                Some(next_month) => next_month,
+                None => return Ok(None),
+            };
+
+            date = match self.find_next_month(date) {
+                Some(date) if date > bound => return Err(OutOfBound),
+                Some(date) => date,
+                None => return Ok(None),
+            };
 
-                return Some(NaiveTime::from_hms(next_hour, minute, 0));
+            match self.find_next_day(date) {
+                Some(next_day) if next_day > bound => return Err(OutOfBound),
+                Some(next_day) => return Ok(Some(next_day)),
+                None => {}
             }
         }
-
-        None
     }
+}
 
-    fn find_next_date(&self, date: Date<Utc>) -> Option<Date<Utc>> {
-        const VALID_NEXT_DAY: &str = "Day should be valid for giving month";
-        const VALID_NEXT_MONTH: &str = "Month should be valid";
+struct OutOfBound;
 
-        let mut month = date.month();
-        if self.months.contains_month(date) {
-            if let Some(next_day) = self.find_next_day(date) {
-                return Some(date.with_day0(next_day).expect(VALID_NEXT_DAY));
-            }
-        }
+#[inline]
+fn minute_floor(dt: DateTime<Utc>) -> DateTime<Utc> {
+    dt.with_second(0)
+        .expect("zero is a valid second value")
```

Comment:
As a developer, I always prefer when my dependencies return Result instead of panicking. That way, I can make a judgement about whether I want my binary to panic when my dependencies has an error. That's just my personal preference. But returning Result always lets the caller decide to panic anyway (or to check what type of error it is, and decide which ones are panic-worthy).

Reply:
If this function panics there's a bug with chrono and I don't think anyone if going to handle a bug with chrono.

Reply:
I understand, but users of this library might prefer to just fail the particular operation/request/workflow they're doing, log an error, and move on to the next one. Rather than bringing down the entire program.

Reply:
Yes but this operation doesn't fail. That's why it's an expect. `with_second` only fails if the second value is out of range (it's not) and if the conversion back to a normal DateTime fails (it can't since the conversion doesn't fail for UTC time zones). Specifically, see:
```rust
    fn offset_from_local_date(&self, _local: &NaiveDate) -> LocalResult<Utc> {
        LocalResult::Single(Utc)
    }
    fn offset_from_local_datetime(&self, _local: &NaiveDateTime) -> LocalResult<Utc> {
        LocalResult::Single(Utc)
    }
```